### PR TITLE
Codechange: add and use the ability to store (lists of) structs in savegames

### PR DESCRIPTION
**Pull request by TrueBrain**
## Motivation / Problem

In a few places (stations, vehicles, towns, companies, ..) the SaveLoad code needs to handle either structs in structs, or lists of structs in structs.

This is all done manually, and not only hard to read, also easy to mess up.

## Description

Introduce `SL_STRUCT` and `SL_STRUCTLIST` to embed a struct in another struct.

For example, for Vehicles it used to have a special `SL` command to embed the common vehicle structure. This meant that SaveLoad needed to know about Vehicles.

Now, the common vehicle sub-struct is called `SlVehicleCommon` and to save it you can use:
```
SLEG_STRUCT(SlVehicleCommon)
```

This works with structs in structs in structs in structs in structs in ... just fine too.

NOTE: I made several commits to introduce `SL_STRUCT` in the rest of the code, as I am hoping that makes reviewing easier. 
NOTE 2: I left a lot of odd constructs in there, just to make review easier. The code really could use a cleanup, but .. I am trying not to do everything at once, also for my own sanity :D
NOTE 3: there should be no difference before or after this PR in respect to what is written/loaded from disk.

## Limitations

- I tested this with ~500 savegames, ranging from TTDp till 1.11.2, validating that they load and run. Nothing that wasn't broken before this PR is broken now. So I think there are no regressions, but with this many changes in the SaveLoad code, that cannot be guaranteed.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/saveload/cheat_sl.cpp [hunk 1]
```diff
@@ -14,37 +14,60 @@
...
  * Load the cheat values.
  */
 static void Load_CHTS()
 {
-	Cheat *cht = (Cheat*)&_cheats;
-	size_t count = SlGetFieldLength() / 2;
+	size_t count = SlGetFieldLength();
 	/* Cannot use lengthof because _cheats is of type Cheats, not Cheat */
-	if (count > sizeof(_cheats) / sizeof(Cheat)) SlErrorCorrupt("Too many cheat values");
+	if (count / 2 > sizeof(_cheats) / sizeof(Cheat)) SlErrorCorrupt("Too many cheat values");
```
Comment (rubidium42) [inline src/saveload/cheat_sl.cpp]: Maybe you could use `count > lengthof(_cheats_desc)`? Then you could also remove the nasty dummies.

Having said that, this commit does not seem to be part/mentioned in this PR. Are you trying to cheat it in? ;)
Or did you want to consider using SLE_STRUCT for each of the Cheats?

File: src/saveload/saveload.h [hunk 2]
```diff
@@ -399,6 +399,27 @@ struct ChunkHandler {
 /** A table of ChunkHandler entries. */
 using ChunkHandlerTable = span<const ChunkHandler>;
 
+/** A table of SaveLoad entries. */
+using SaveLoadTable = span<const struct SaveLoad>;
+
+class SaveLoadHandler {
```
Comment (rubidium42) [inline src/saveload/saveload.h:407]: I guess the draftiness of this PR means that the comments here are still pending / that you're waiting on glx's work, right?

File: src/saveload/town_sl.cpp [hunk 3]
```diff
@@ -113,7 +113,89 @@ void UpdateHousesAndTowns()
...
+			SlObject(&t->received[i], this->GetDescription());
+		}
+	}
+};
+
+class SlTownAcceptanceMatrix : public SaveLoadHandler {
+public:
+	SaveLoadTable GetDescription()
+	{
+		/* Here due to private member vars. */
```
Comment (rubidium42) [inline src/saveload/town_sl.cpp]: Is this still needed? I can't see `GetTileMatrixDesc` being made `friend` anywhere, so it seems to be not needed anymore.

File: src/saveload/saveload.h [hunk 4]
```diff
@@ -692,11 +690,8 @@ struct SaveLoad {
  */
 #define SLE_CONDNULL(length, from, to) {SL_ARR, SLE_FILE_U8 | SLE_VAR_NULL, length, from, to, 0, nullptr, 0, nullptr}
 
-/** Translate values ingame to different values in the savegame and vv. */
-#define SLE_WRITEBYTE(base, variable) SLE_GENERAL(SL_WRITEBYTE, base, variable, 0, 0, SL_MIN_VERSION, SL_MAX_VERSION, 0)
-
-#define SLE_VEH_INCLUDE() {SL_VEH_INCLUDE, 0, 0, SL_MIN_VERSION, SL_MAX_VERSION, 0, [] (void *b, size_t) { return b; }, 0}
-#define SLE_ST_INCLUDE() {SL_ST_INCLUDE, 0, 0, SL_MIN_VERSION, SL_MAX_VERSION, 0, [] (void *b, size_t) { return b; }, 0}
+/** Only write byte during saving; never read it during loading. */
```
Comment (rubidium42) [inline src/saveload/saveload.h]: The comment is misleading as the byte is read, just separately using `SlReadByte` to ensure the right (sub)class is constructed. Now it reads like: "we write a byte to the save game, that we never need", and I start wondering. Why write a byte to the save game that is never read?

File: src/saveload/cheat_sl.cpp (hunk 1 above)
Comment (TrueBrain) [reply src/saveload/cheat_sl.cpp]: This commit message making this change literal is called: `Codechange: rework CHTS-chunk save/load to be more like the others` :P I have to see if I have enough work to warrant another "mixed-fixes-for-stupid-code" PR .. but that is also why this PR is still a draft :)

Either way, this was existing code, I really did not parse yet what it said or tries to say. Just moved a line of code from one above to the one below.

What do you mean with "nasty dummies"? What "nasty" and what "dummies" is that referring to?

File: src/saveload/saveload.h (hunk 2 above)
Comment (TrueBrain) [reply src/saveload/saveload.h:407]: it means a lot of things, but most that this is not the final form yet, and that I didn't do a pass myself on any coding guideliness :) I mostly created the PR so I could collect opinions on the way of doing stuff.

File: src/saveload/saveload.h (hunk 4 above)
Comment (TrueBrain) [reply src/saveload/saveload.h]: I am happy to read that you read it like that, as that is exactly what it does. And sure, we have an intended usage of it, but that is just that: how we intend someone to use it .. but nothing is enforcing that.

It really is a shitty mechanism that is very easy to misunderstand and fuck up. So I am more looking into how to remove it completely ;)

That said, I can add a line of doxygen comment to explain the intended use of it .. but at least you now understood correctly what it did :) The initial comment was truly misleading ;)

File: src/saveload/cheat_sl.cpp (hunk 1 above)
Comment (rubidium42) [reply src/saveload/cheat_sl.cpp]: For the dummies, take a look at: https://github.com/OpenTTD/OpenTTD/blob/1e564b333fa2d7bde2f21d7a05bb3e81305185c5/src/cheat_type.h#L31 further on line 33 and 36. That, together with the comment around line 23 made them nasty to me.

File: src/saveload/vehicle_sl.cpp [hunk 5]
```diff
@@ -812,66 +878,98 @@ SaveLoadTable GetVehicleDescription(VehicleType vt)
...
-		_roadveh_desc,
-		_ship_desc,
-		_aircraft_desc,
-		_special_desc,
-		_disaster_desc,
-		_common_veh_desc,
-	};
+	void Save(Vehicle *v) const override { this->GenericSaveLoad(v); }
+	void Load(Vehicle *v) const override { this->GenericSaveLoad(v); }
+	void FixPointers(Vehicle *v) const override { this->GenericSaveLoad(v); }
```
Comment (rubidium42) [inline src/saveload/vehicle_sl.cpp:953]: I'm seeing these three return over and over again, might it be worth it to have an intermediate "super" class with virtual GenericSaveLoad, or maybe even a templated intermediate super class that requires the implementing sub class to provide a GenericSaveLoad? The same is already done with description, and it means less duplicating of these three lines all over the save load code.

File: src/saveload/saveload.h [hunk 6]
```diff
@@ -400,6 +400,68 @@ struct ChunkHandler {
...
+	virtual SaveLoadTable GetDescription() const = 0;
+};
+
+/**
+ * Default handler for saving/loading an object to/from disk.
+ *
+ * This handles a few common things for handlers, meaning the actual handler
+ * needs less code.
+ */
+template <class TOwn, class TObject>
```
Comment (rubidium42) [inline src/saveload/saveload.h]: Doxygen misses for these parameters.

Comment (rubidium42) [inline src/saveload/saveload.h]: TOwn -> TInstance or TImpl? Now it's a bit confusing w.r.t. Towns.

File: src/saveload/vehicle_sl.cpp (hunk 5 above)
Comment (TrueBrain) [reply src/saveload/vehicle_sl.cpp:953]: Later PRs mostly remove this; it is simply as a stop-gap. As I knew if I would to copy/paste the code and create identical functions I would get the complaints I have duplicated code :P This is the best I could do in the chain of PRs :)  Adding an abstraction now would just be a waste of my time, basically, as I would remove it 1 PR later :P I am lazy :D

But even after all this work there is plenty of room to refactor the code to have less code. But I left this out of this set of PRs, as reviewing this is already difficult enough :)

(just to confirm for myself, after #9322:
```
$ grep -R -n GenericSaveLoad ../src/ | wc -l
0
```
)

File: src/saveload/station_sl.cpp [hunk 7]
```diff
@@ -374,89 +374,118 @@ static void Ptrs_STNS()
...
+		SlObject(bst, this->GetDescription());
+	}
+
+	void Save(BaseStation *bst) const override { this->GenericSaveLoad(bst); }
+	void Load(BaseStation *bst) const override { this->GenericSaveLoad(bst); }
+	void FixPointers(BaseStation *bst) const override { this->GenericSaveLoad(bst); }
 };
 
-static OldPersistentStorage _old_st_persistent_storage;
+class SlStationCommon : public DefaultSaveLoadHandler<SlStationCommon, BaseStation> {
```
Comment (rubidium42) [inline src/saveload/station_sl.cpp]: Maybe SlNormalStation as now it sort-of implies there might be a sub class?

File: src/saveload/station_sl.cpp [hunk 8]
```diff
@@ -374,89 +374,118 @@ static void Ptrs_STNS()
...
-static const SaveLoad _base_station_desc[] = {
-	      SLE_VAR(BaseStation, xy,                     SLE_UINT32),
-	      SLE_REF(BaseStation, town,                   REF_TOWN),
-	      SLE_VAR(BaseStation, string_id,              SLE_STRINGID),
-	     SLE_SSTR(BaseStation, name,                   SLE_STR | SLF_ALLOW_CONTROL),
-	      SLE_VAR(BaseStation, delete_ctr,             SLE_UINT8),
-	      SLE_VAR(BaseStation, owner,                  SLE_UINT8),
-	      SLE_VAR(BaseStation, facilities,             SLE_UINT8),
-	      SLE_VAR(BaseStation, build_date,             SLE_INT32),
+class SlStationBase : public DefaultSaveLoadHandler<SlStationBase, BaseStation> {
```
Comment (rubidium42) [inline src/saveload/station_sl.cpp:508]: I'd note in the comment for this one that SlWaypoint also makes use of this.

**Review by rubidium42 — APPROVED** [review]

**Pull request merged**